lib/translator/read.cc

```diff
@@ -147,19 +139,22 @@ StatusCode Read6ToNvme(Span<const uint8_t> scsi_cmd,
   uint16_t updated_transfer_length =
       read_cmd.transfer_length == 0 ? 256 : read_cmd.transfer_length;
 
-  // Transform logical_block_address to network endian before casting to 64 bits
   StatusCode status = LegacyRead(
-      htonl(read_cmd.logical_block_address), nvme_cmd, allocation, nsid,
+      nvme_cmd, allocation, nsid,
       GetTransferLengthPages(updated_transfer_length, page_size, lba_size));
-
   if (status != StatusCode::kSuccess) {
     return status;
   }
+  // cdw10 Starting lba bits 31:00
+  uint32_t host_endian_lba = (read_cmd.logical_block_address_1 << 16) +
+                             ntohs(read_cmd.logical_block_address_2);
+  nvme_cmd.cdw[0] = __bswap_32(htonl(host_endian_lba));
 
-  // cdw12 nlb bits 15:00 (zero based field)
-  nvme_cmd.cdw[2] = updated_transfer_length - 1;
+  // nlb is a zero-based field
+  // cdw12 nlb bits 15:00
+  nvme_cmd.cdw[2] = __bswap_16(htons(updated_transfer_length - 1));
```

Comment:
Is cdw[] defined as 32 or 16 bit?  Assigning a 16 bit value to a 32 bit variable will have the compiler zero-pad the value based on host byte order.

Reply:
good catch, it is 32 bit